src/fsharp/FSharp.Core/seq.fs

```diff
@@ -842,6 +812,397 @@ namespace Microsoft.FSharp.Collections
+
+                override second.ComposeFilter (first:Filter<'T>) : SeqComponent<'T,'T> =
+                    upcast Filter (Helpers.ComposeFilter first.Filter second.Filter)
+
+                override second.ComposeMapFilter<'S> (first:MapFilter<'S,'T>) : SeqComponent<'S,'T> =
+                    upcast MapFilter (first.Map, Helpers.ComposeFilter first.Filter second.Filter)
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'T>) : bool = 
+                    if filter input then
+                        output <- input
+                        true
+                    else
+                        false
+
+                member __.Filter :'T->bool = filter
+
+            and MapFilter<'T,'U> (map:'T->'U, filter:'U->bool) =
+                inherit SeqComponent<'T,'U>()
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'U>) :bool = 
+                    output <- map input
+                    Helpers.avoidTailCall (filter output)
+
+                override first.Composer (second:SeqComponent<'U,'V>):SeqComponent<'T,'V> =
+                    second.ComposeMapFilter first
+
+                member __.Map    : 'T->'U = map
+                member __.Filter : 'U->bool = filter
+
+            and FilterMap<'T,'U> (filter:'T->bool, map:'T->'U) =
+                inherit SeqComponent<'T,'U>()
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'U>) : bool = 
+                    if filter input then
+                        output <- map input
+                        true
+                    else
+                        false
+
+                override this.Composer (next:SeqComponent<'U,'V>) : SeqComponent<'T,'V> = next.ComposeFilterMap this
+
+                member __.Filter : 'T->bool = filter
+                member __.Map    : 'T->'U = map
+
+            and Pairwise<'T> () =
+                inherit SeqComponent<'T,'T*'T>()
+
+                let mutable isFirst = true
+                let mutable lastValue = Unchecked.defaultof<'T>
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'T*'T>) : bool = 
+                    if isFirst then
+                        lastValue <- input
+                        isFirst <- false
+                        false
+                    else
+                        output <- lastValue, input
+                        lastValue <- input
+                        true
+
+            and Skip<'T> (skipCount:int) =
+                inherit SeqComponent<'T,'T>()
+
+                let mutable count = 0
+
+                override first.Composer (second:SeqComponent<'T,'V>) : SeqComponent<'T,'V> =
+                    second.ComposeSkip first
+
+                override second.ComposeSkip (first:Skip<'T>) : SeqComponent<'T,'T> =
+                    if System.Int32.MaxValue - second.SkipCount > first.SkipCount then
```

Comment:
Seems like this condition would make more sense as `Int32.MaxValue > first.SkipCount + second.SkipCount` to match expression in the next line.